src/test/java/codeu/controller/RegisterServletTest.java

```diff
@@ -5,30 +5,84 @@
   private HttpServletResponse mockResponse;
   private RequestDispatcher mockRequestDispatcher;
 
- @Before
+  @Before
   public void setup() {
     registerServlet = new RegisterServlet();
     mockRequest = Mockito.mock(HttpServletRequest.class);
     mockResponse = Mockito.mock(HttpServletResponse.class);
     mockRequestDispatcher = Mockito.mock(RequestDispatcher.class);
     Mockito.when(mockRequest.getRequestDispatcher("/WEB-INF/view/register.jsp"))
-       .thenReturn(mockRequestDispatcher);
+        .thenReturn(mockRequestDispatcher);
   }
 
- @Test
+  @Test
   public void testDoGet() throws IOException, ServletException {
     registerServlet.doGet(mockRequest, mockResponse);
     Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
   }
+
+  @Test
+  public void testDoPost_BadUsername() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("bad !@#$% username");
+    registerServlet.doPost(mockRequest, mockResponse);
+    Mockito.verify(mockRequest)
+        .setAttribute("error", "Please enter only letters, numbers, and spaces.");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testDoPost_NewUser() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("test username");
+    Mockito.when(mockRequest.getParameter("password")).thenReturn("abcdtest");
+
+    UserStore mockUserStore = Mockito.mock(UserStore.class);
+    Mockito.when(mockUserStore.isUserRegistered("test username")).thenReturn(false);
+    registerServlet.setUserStore(mockUserStore);
+
+    HttpSession mockSession = Mockito.mock(HttpSession.class);
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
+
+    registerServlet.doPost(mockRequest, mockResponse);
+
+    ArgumentCaptor<User> userArgumentCaptor = ArgumentCaptor.forClass(User.class);
+
+    Mockito.verify(mockUserStore).addUser(userArgumentCaptor.capture());
+    Assert.assertEquals(userArgumentCaptor.getValue().getName(), "test username");
+    Assert.assertEquals(userArgumentCaptor.getValue().getPassword(), "abcdtest");
+    Mockito.verify(mockResponse).sendRedirect("/login");
+  }
+
+  @Test
+  public void testDoPost_IsRegistered() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("test username");
+
+    UserStore mockUserStore = Mockito.mock(UserStore.class);
+    Mockito.when(mockUserStore.isUserRegistered("test username")).thenReturn(true);
+    registerServlet.setUserStore(mockUserStore);
+
+    HttpSession mockSession = Mockito.mock(HttpSession.class);
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
+
+    registerServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verify(mockUserStore, Mockito.never()).addUser(Mockito.any(User.class));
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
```

Comment:
I see that LoginServletTest.java has the following, instead of your 86
    Mockito.verify(mockSession).setAttribute("user", "test username");
    Mockito.verify(mockResponse).sendRedirect("/conversations");

Reply:
Thanks for catching that! I checked the code and since there is some difference between LoginServlet and RegisterServlet so test cases vary a little here. But I double checked and it is reasonable to set the username text the account the user register, I will follow up on that!